### ISSUE TITLE
Running model on Android smartphone

### ISSUE THREAD

**Issue by gilbeckers** [ildoonet/tf-pose-estimation#110]
Is it possible to deploy the TF model on Android? Did you (or someone else) tried to run this on a smartphone before?

If not, what would be the steps to make this possible? I guess port the intercation between python en TF to java.. ? 

Thanks in advance!!

**Comment by ouceduxzk** [comment]
it's definitely possible, check the usage of tensorflow-lite

**Comment by gilbeckers** [comment]
Yeah I successfully deployed the object detection demo (inception) on Android. However, I was wondering which steps are needed in order to deploy this particular model (mobilenet version)

**Comment by ouceduxzk** [comment]
well, if you know how to deploy one tf model, when why this particular model is different, mobilenet version is no difference , the layers are supported in tf, nothing stop you doing the exact same steps you did for inception

**Comment by gilbeckers** [comment]
Wow, that would be great! 
For the image extraction, I suggest you take a look at the Object Detection Demo of Tensorflow Android. 
This helped me a lot, both for starting with Tensorflow on Android and image processing on Android in general

**Comment by jax79sg** [comment]
Hi, thanks for the Tensorflow Android tip. Been using that as my base before moving on to the SNPE SDK. I'm facing an issue though. What kind of output am i expecting from the Openpose/concat_stage7 node? I am getting bufferoverflow exceptions everytime i run the inference. For example, a typical classifier would have the softmax layer and output a single float. It shows the following on tensorboard.
`
Attributes (3)

N{"i":2}
T{"type":"DT_FLOAT"}
Tidx{"type":"DT_INT32"}
`

**Comment by ouceduxzk** [comment]
@jax79sg  can you post the entire error msg, it is hard to tell what happens. Regarding to the output from the network, it is 46x54x38 and 46x54x19, think about the problem, this is not classification or detection, the output is 3-dim tensor

**Comment by jax79sg** [comment]
Hi @ouceduxzk , thanks for the response. I have raised a separate [issue](https://github.com/ildoonet/tf-pose-estimation/issues/120) earlier. I'm having problems understanding the 46x54x38 and 46x54x19 output, would you please point me to the documentation on these tensors?

Thank you.

**Comment by ouceduxzk** [comment]
@jax79sg  please read the original paper of cmu pose, https://arxiv.org/abs/1611.08050, the 38 represents 19 pairs of PAF, 19 represents 18 keypoints + 1 background, 46 or 54 is the height or width/8, 8 means your network downsample 3 times, each time the size is two times smaller. Please understand the paper first before you want to experiments with the code, it also helps you to understand some other cv problems, like semantic segmentation, object detection, and so on.

**Comment by jax79sg** [comment]
Hi @ouceduxzk , thanks. I understood from the paper that there were 2 'sub-outputs' after which they are concat in the network. A confidence map of size WxHxPartX where each be coded as  (x,y,Part1)->Confidence .... (x,y,Part19) ->Confidence. Effectively if i can ensure only 1 person in the scene all the time, i can make do with this half of the network without going into PAF.

The second half of the network is where i got a bit lost. The PAF describes the limb (Part to Part) and encodes it with a 2d vector that indicates the direction. I couldn't really get my head around how it looks like in code. WxHxPartX where each be coded as (x,y,Part1)-> (Part2,(x,y)) .... (x,y,Part19) -> (Part19,(x,y)) ?? 

Thanks

**Comment by GajjarMihir** [comment]
Hello, 

I also aim to run pose estimation on android and for the same, I followed the object detection in android tutorial suggested by @gilbeckers. Can anyone help me to implement pose estimation, in the same way, using the .pb file given?

I greatly appreciate your help.

**Comment by zyxcambridge** [comment]
I managed to loan a Galaxy S8 ( With Qualcomm's Snapdragon 835 ) for a week and i thought of implementing tf-openpose on Android using Snapdragon Neural Processing Engine SDK 


能够达到多少fps 来着？

**Comment by GajjarMihir** [comment]
Hello @zyxcambridge, were you able to port it to android? If yes, can you please help me with this?

**Comment by GajjarMihir** [comment]
@jax79sg can you please help me with this? I am also trying to port tf-openpose on android based devices.

**Comment by GajjarMihir** [comment]
I figured out that opencv 3.4.1 has inbuilt library for openpose. I was able to get prediction for an image on ubuntu using opencv 3.4.1. However, I am getting errors while implementing it on android. It would be really helpful to me if anybody has managed to successfully implement openpose on android.

**Comment by GajjarMihir** [comment]
@jax79sg @zyxcambridge  In java in what type and size of variable did you store the output of the model. I am facing the same issue.

**Comment by GajjarMihir** [comment]
I stored the output of "Openpose/concat_stage7" node into a float variable of size 46*46*57 as the size of my image is 368*368. I am getting the output but I dont know how to extract the coordinates from this float variable. If anyone can help to get the coordinates from this float array then it would be really helpful.

**Comment by GajjarMihir** [comment]
@gilbeckers @ouceduxzk @jax79sg @ildoonet @zyxcambridge 

I stored the output of "Openpose/concat_stage7" node into a float variable of size 46×46×57 as the size of my image is 368*368. I am getting the output but I don't know how to extract the coordinates from this float variable. 

I extracted the coordinates in the python code as follows:
'''
for human in humans:
	# draw point
	    for i in range(common.CocoPart.Background.value):
	        if i not in human.body_parts.keys():
	       	    continue

	       	print(human.body_parts[i])
'''

If anyone can help to get the coordinates from this float array in java then it would be really helpful.

**Comment by ouceduxzk** [comment]
@GajjarMihir  This repo has the python code of connecting limbs and get the coordinates, why not just translate it into java code line by line in the worst case ?  To understand the logic, please read the original paper

**Comment by GajjarMihir** [comment]
ok. I will try that. Thanks :)

**Comment by jpizarrom** [comment]
https://github.com/ildoonet/tf-pose-estimation/issues/2#issuecomment-388634576

**Comment by lyyiangang** [comment]
@jax79sg  I am also trying openpose on NPE. Hope we can something to exchange.

**Comment by jax79sg** [comment]
Hi @lyyiangang , i did try this on a Xperia XZ2 but ran out of time to fully work on it. I extracted the gist here, you may try it out and let me know if it works on your Qualcomm chip.
[https://gist.github.com/jax79sg/f111bfbb6d1ef87e41cbacdfe70d8a7e](https://gist.github.com/jax79sg/f111bfbb6d1ef87e41cbacdfe70d8a7e)

**Comment by liabixiaoxin** [comment]
@jax79sg 
Personal tips: if you find it difficult debugging directly with apk, you'd better use not do it this way since you have to write extra Java code. 
Using `adb shell` to login a development board or SoC with Qualcomm chip is a better solution. 

like this one: [Open-Q™ 835 HDK Development Kit](https://shop.intrinsyc.com/collections/product-development-kits/products/open-q-835-development-kit)

**Comment by llhe** [comment]
To speedup the inference speed with mobile GPU, you can checkout this project: https://github.com/xiaomi/mace, and here is some benchmark results: https://github.com/XiaoMi/mace/issues/1 and also benchmark results with other frameworks (including SNPE): https://github.com/XiaoMi/mobile-ai-bench

**Comment by smartether** [comment]
InvalidArgumentError (see above for traceback): Restoring from checkpoint failed. This is most likely due to a mismatch between the current graph and the graph from the checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Assign requires shapes of both tensors to match. lhs shape= [3,3,48,1] rhs shape= [3,3,64,1]
         [[{{node save/Assign_378}} = Assign[T=DT_FLOAT, _class=["loc:@MobilenetV1/Conv2d_2_depthwise/depthwise_weights"], use_locking=true, validate_shape=true, _device="/job:localhost/replica:0/task:0/device:CPU:0"](MobilenetV1/Conv2d_2_depthwise/depthwise_weights, save/RestoreV2_359:19)]]

this error raise when convert your model to tensorflow-lite format with graph.pd and mobilenet_v1_1.0_224_2017_06_14.ckpt

**Comment by bsherifi** [comment]
@ankitksharma I am trying to convert to tflite too, so if you have progressed let me know.

**Comment by vidheyoza** [comment]
Did you manage to create the graph.pb file for the apk from https://github.com/edvardHua/PoseEstimationForMobile ?

**Comment by bsherifi** [comment]
Yes I always achieved to create the tflite file which is supposed to run on Android. However, it always crashed. Something wrong with the conversion process I guess.

**Comment by vidheyoza** [comment]
So the openpose output is of 18 keypoints if I'm not mistaken, which the app I referred above has designed above is developed for a 14 keypoint data. 

Is that the problem you are facing? Or are you not able to make the .pb file itself?

**Comment by ankitksharma** [comment]
@vidheyoza Were you able to run this model using `tflite` on Android or python? If yes, then could you check the code snippet I shared in https://github.com/ildoonet/tf-pose-estimation/issues/110#issuecomment-577180936 and let me know what I missed

**Comment by MadisLemsalu** [comment]
Curious to know as well if someone managed to get it to work and how
